### ISSUE TITLE
mlflow + pytorch lightning not able to log custom models

### ISSUE THREAD

**Issue by piseabhijeet** [mlflow/mlflow#4763]
Thank you for submitting an issue. Please refer to our [issue policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md) for additional information about bug reports. For help with debugging your code, please refer to [Stack Overflow](https://stackoverflow.com/questions/tagged/mlflow).

**Please fill in this bug report template to ensure a timely and thorough response.**

### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [ ] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ -] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **MLflow installed from (source or binary)**:
- **MLflow version (run ``mlflow --version``)**:
- **Python version**:
- **npm version, if running the dev UI**:
- **Exact command to reproduce**:

### Describe the problem
Describe the problem clearly here. Include descriptions of the expected behavior and the actual behavior.

### Code to reproduce issue
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

### Other info / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

### What component(s), interfaces, languages, and integrations does this bug affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ -] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [- ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

I am using a custom pytorch lightning trainer with  mlflow logger
_**MLFlowLogger(experiment_name="Experiment", tracking_uri=tracking_uri)**_

**_I use both mlflow.pytorch.autolog()_**

and the logger - which saves the parameters in default run and metrics in the run in "Experiment". However it does not log the model which i want to use for serving, it gives the following error:

![image](https://user-images.githubusercontent.com/57705684/131857341-2cc063c2-e9bb-4f16-930d-ae8c47205915.png)

Is it that pltrainer models are not supported? My model has  single inputs and multiple outputs.

**Comment by jwyyy** [comment]
You should use a `nn.Module` (`self.model` if it is a pytorch model) in the first argument not a state dict (`self.model.state_dict()`). The check occurs [here](https://github.com/mlflow/mlflow/blob/master/mlflow/pytorch/__init__.py#L460). Please also read [this line](https://github.com/mlflow/mlflow/blob/master/mlflow/pytorch/__init__.py#L133).

**Comment by piseabhijeet** [comment]
> You should use a `nn.Module` (`self.model` if it is a pytorch model) in the first argument not a state dict (`self.model.state_dict()`). The check occurs [here](https://github.com/mlflow/mlflow/blob/master/mlflow/pytorch/__init__.py#L460). Please also read [this line](https://github.com/mlflow/mlflow/blob/master/mlflow/pytorch/__init__.py#L133).

Hi,

I tried using `self.model` as well but it throws me this error:
![image](https://user-images.githubusercontent.com/57705684/133026005-39403106-38b4-4fe0-beec-3e9e00e3583a.png)

Hence i cannot serve my model using mlflow model serving. Is there a better alternative?

**Comment by jwyyy** [comment]
Can you provide more code? I am not able to identity the problem based on this screenshot. You can find some posts related to the error: [link1](https://github.com/kubeflow/pipelines/issues/4969) and [link2](https://github.com/tensorflow/tensorflow/issues/41159).

**Comment by piseabhijeet** [comment]
@jwyyy 
Thank you for the prompt response.

Unfortunately, I cannot put the entire code/ traceback here due to some confidentiality agreements. However, i can provide you with the flow we have:

I am using tfrecords and tf.data pipeline for our pytorch models. Hence while saving the entire model, i am guessing Tensorflow errors pop up.  The entire stack trace is here:

![image](https://user-images.githubusercontent.com/57705684/133029951-2035c796-f80a-4082-92f9-36d5110921ff.png)

As a workaround, i tried saving the model's 'state_dict()' instead of the entire model but it fails.

**Comment by piseabhijeet** [comment]
@jwyyy
Thank you for the prompt response.

Unfortunately, I cannot put the entire code/ traceback here due to some confidentiality agreements. However, i can provide you with the flow we have:

I am using Tensorflow and Pytorch at different stages in the model training pipeline. I am assuming that while saving the entire Pytorch model, the dependency on Tensorflow could be causing the problem. The entire stack trace is here:

![image](https://user-images.githubusercontent.com/57705684/133033874-e7d54406-bfe1-4c50-8d8c-7fce507ce84a.png)


As a workaround, i tried saving the model's 'state_dict()' instead of the entire model but it fails.

**Comment by jwyyy** [comment]
Ok, that's fine. Is there any specific reason that you want to use two backends? I guess this caused the error?

**Comment by piseabhijeet** [comment]
Yes, i want to make tensorflow-pytorch compatible training pipeline.
